src/App.tsx

```diff
@@ -1,26 +1,234 @@
+
+  useEffect(() => {
+    if (!errorMessage) {
+      return;
+    }
+
+    const timerID = setTimeout(() => {
+      setErrorMessage('');
+    }, 3000);
+
+    return () => clearTimeout(timerID);
+  }, [errorMessage]);
+
+  useEffect(() => {
+    if (!temptTodo) {
+      inputRef.current?.focus();
+    }
+  }, [temptTodo]);
+
+  const setParams = (todoId: number) => {
+    setErrorMessage('');
+    setProcessingIds(prevIds => [...prevIds, todoId]);
+  };
+
+  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    const normalizedTitle = title.trim();
+
+    if (!normalizedTitle) {
+      setErrorMessage('Title should not be empty');
+
+      return;
+    }
+
+    setTemptTodo({
+      id: 0,
+      title: normalizedTitle,
+      completed: false,
+      userId: 0,
+    });
+
+    setErrorMessage('');
+
+    createTodo(normalizedTitle)
+      .then(todo => {
+        setTodos((prevTodos: Todo[]) => [...prevTodos, todo]);
+        setTitle('');
+      })
+      .catch(() => setErrorMessage('Unable to add a todo'))
+      .finally(() => setTemptTodo(null));
+  };
+
+  const handleDeletingTodo = (todoIdDelete: number) => {
+    setParams(todoIdDelete);
+
+    deleteTodo(todoIdDelete)
+      .then(() => {
+        setTodos(prevTodos =>
+          prevTodos.filter(todo => todo.id !== todoIdDelete),
+        );
+      })
+      .catch(() => setErrorMessage('Unable to delete a todo'))
+      .finally(() => {
+        setProcessingIds(prevIds => prevIds.filter(id => id !== todoIdDelete));
+        inputRef.current?.focus();
+      });
+  };
+
+  const handleClearComplitedTodos = () => {
+    completedTodos.forEach(todo => handleDeletingTodo(todo.id));
+  };
+
+  const toggleTodo = (todoToUpdate: Todo) => {
+    setParams(todoToUpdate.id);
+
+    updateTodo({
+      ...todoToUpdate,
+      completed: !todoToUpdate.completed,
+    })
+      .then(updatedTodo => {
+        setTodos(prevTodos =>
+          prevTodos.map(todo =>
+            todo.id === updatedTodo.id ? updatedTodo : todo,
+          ),
+        );
+      })
+      .catch(() => setErrorMessage('Unable to update a todo'))
+      .finally(() => {
+        setProcessingIds(prevIds => {
+          return prevIds.filter(id => id !== todoToUpdate.id);
+        });
+      });
+  };
+
+  const toggloAll = () => {
+    if (activeTodos.length !== 0) {
+      activeTodos.forEach(todo => toggleTodo(todo));
+    } else {
+      completedTodos.forEach(todo => toggleTodo(todo));
+    }
+  };
+
+  const renameTodo = (todoRename: Todo, newTitle: string) => {
+    setParams(todoRename.id);
+
+    return updateTodo({
+      ...todoRename,
+      title: newTitle,
+    })
+      .then(updatedTodo => {
+        setTodos(prevTodos =>
+          prevTodos.map(todo =>
+            todo.id === updatedTodo.id ? updatedTodo : todo,
+          ),
+        );
+      })
+      .catch(error => {
+        setErrorMessage('Unable to update a todo');
+        throw error;
```

Comment:
You don't need to throw error, if you don't catch and manage it in the next step